velox/functions/prestosql/aggregates/ChecksumAggregate.cpp

```diff
@@ -0,0 +1,235 @@
+#include "velox/functions/prestosql/aggregates/AggregateNames.h"
+#include "velox/functions/prestosql/aggregates/PrestoHasher.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::aggregate {
+
+namespace {
+
+/// Computes an order-insensitive checksum of the input
+/// vector.
+///
+/// checksum(T)-> varbinary
+class ChecksumAggregate : public exec::Aggregate {
+ public:
+  explicit ChecksumAggregate(const TypePtr& resultType)
+      : Aggregate(resultType) {}
+
+  void finalize(char** /* groups */, int32_t /* numGroups */) override {}
+
+  int32_t accumulatorFixedWidthSize() const override {
+    return sizeof(int64_t);
+  }
+
+  void initializeNewGroups(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    setAllNulls(groups, indices);
+    for (auto i : indices) {
+      *value<int64_t>(groups[i]) = 0;
+    }
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    auto* vector = (*result)->asUnchecked<FlatVector<StringView>>();
+    VELOX_CHECK(vector);
+    vector->resize(numGroups);
+
+    auto* rawValues = vector->mutableRawValues();
+    for (auto i = 0; i < numGroups; ++i) {
+      auto group = groups[i];
+      if (isNull(group)) {
+        vector->setNull(i, true);
+      } else {
+        vector->setNull(i, false);
+        rawValues[i] = StringView(value<char>(group), sizeof(int64_t));
+      }
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    auto vector = (*result)->asUnchecked<FlatVector<int64_t>>();
+    VELOX_CHECK(vector);
+    vector->resize(numGroups);
+
+    auto* rawValues = vector->mutableRawValues();
+    vector->clearAllNulls();
+    for (auto i = 0; i < numGroups; ++i) {
+      auto group = groups[i];
+      if (isNull(group)) {
+        vector->setNull(i, true);
+        continue;
+      }
+      rawValues[i] = *value<int64_t>(group);
+    }
+  }
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushDown*/) override {
+    const auto& arg = args[0];
+    auto hasher = getPrestoHasher(arg->type());
+    auto hashes = getHashBuffer(rows.end(), arg->pool());
+    hasher->hash(arg, rows, hashes);
+    auto rawHashes = hashes->as<int64_t>();
+
+    rows.applyToSelected([&](vector_size_t row) {
+      clearNull(groups[row]);
+      if (arg->isNullAt(row)) {
+        computeHashForNull(groups[row]);
+      } else {
+        computeHash(groups[row], rawHashes[row]);
+      }
+    });
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushDown*/) override {
+    const auto& arg = args[0];
+    auto vector = arg->asUnchecked<FlatVector<int64_t>>();
+    auto rawValues = vector->rawValues();
+
+    rows.applyToSelected([&](vector_size_t row) {
```

Comment:
Perhaps, add a comment explaining why null check is missing here. Also, since these functions can be used outside of Presto, e.g. by XStream, it might be safer to not assume that there will be no nulls in the intermediate results and add a check.

Reply:
In the case of checksum,  I dont see how it should be  possible for intermediate groups to have null value . However I have added a null check.